File: src/test/java/nl/rabobank/kotlinmovement/recipes/test/RecipeMockMvcTest.java
package nl.rabobank.kotlinmovement.recipes.test;

import com.fasterxml.jackson.databind.ObjectMapper;
import nl.rabobank.kotlinmovement.recipes.test.model.IngredientRequestTest;
import nl.rabobank.kotlinmovement.recipes.test.model.RecipeRequestTest;
import nl.rabobank.kotlinmovement.recipes.test.model.RecipeResponseTest;
import nl.rabobank.kotlinmovement.recipes.test.model.RecipesErrorResponseTest;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.annotation.DirtiesContext;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.ResultMatcher;
import org.springframework.test.web.servlet.request.MockHttpServletRequestBuilder;

import java.util.Set;

import static nl.rabobank.kotlinmovement.recipes.test.RecipeTestData.getDefaultIngredientRequests;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.put;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@SpringBootTest
@AutoConfigureMockMvc
@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
public class RecipeMockMvcTest {
    @Autowired
    protected MockMvc mockMvc;
    protected static ObjectMapper objectMapper = new ObjectMapper();

    protected RecipeRequestTest initRecipeRequest;

    protected void setInitialState() throws Exception {
        final Set<IngredientRequestTest> ingredients = getDefaultIngredientRequests();
        RecipeRequestTest initRecipe = new RecipeRequestTest("Pizza", ingredients);
        initRecipeRequest = initRecipe;
        createRecipe(initRecipe);
    }

    protected <T> T mockMvcPerformRequest(MockHttpServletRequestBuilder requestBuilder, Class<T> responseType, ResultMatcher status) throws Exception {
        var responseArrayString = mockMvc
                .perform(requestBuilder)
                .andDo(print())
                .andExpect(status)
                .andReturn()
                .getResponse()
                .getContentAsString();
        return objectMapper.readValue(responseArrayString, responseType);
    }

    protected void mockMvcPerformRequest(MockHttpServletRequestBuilder requestBuilder, ResultMatcher status) throws Exception {
        mockMvc
                .perform(requestBuilder)
                .andDo(print())
                .andExpect(status)
                .andReturn()
                .getResponse()
                .getContentAsString();
    }

    protected RecipeResponseTest getRecipe(long id) throws Exception {
        return mockMvcPerformRequest(get("/recipes/{id}", id), RecipeResponseTest.class, status().isOk());
    }

    protected RecipeResponseTest[] getAllRecipes() throws Exception {
        return mockMvcPerformRequest(get("/recipes"), RecipeResponseTest[].class, status().isOk());
    }


    protected RecipeResponseTest updateRecipe(Long id, RecipeRequestTest recipeRequest) throws Exception {
        final MockHttpServletRequestBuilder requestBuilder =
                put("/recipes/{id}", id)
                        .content(objectMapper.writeValueAsString(recipeRequest))
                        .contentType(MediaType.APPLICATION_JSON);
        return mockMvcPerformRequest(requestBuilder, RecipeResponseTest.class, status().isOk());
    }

    protected RecipesErrorResponseTest badRequestCall(MockHttpServletRequestBuilder builder) throws Exception {
        final MockHttpServletRequestBuilder requestBuilder =
                builder.contentType(MediaType.APPLICATION_JSON);
        return mockMvcPerformRequest(requestBuilder, RecipesErrorResponseTest.class, status().isBadRequest());
    }

    protected RecipesErrorResponseTest notFoundCall(MockHttpServletRequestBuilder builder) throws Exception {
        final MockHttpServletRequestBuilder requestBuilder =
                builder.contentType(MediaType.APPLICATION_JSON);
        return mockMvcPerformRequest(requestBuilder, RecipesErrorResponseTest.class, status().isNotFound());
    }

    private void createRecipe(RecipeRequestTest recipe) throws Exception {
        mockMvcPerformRequest(
                post("/recipes")
                        .content(objectMapper.writeValueAsString(recipe))
                        .contentType(MediaType.APPLICATION_JSON)
                , RecipeResponseTest.class, status().isCreated());
    }

}

File: recipe-kotlin-non-blocking/src/test/kotlin/nl/rabobank/kotlinmovement/recipes/test/util/RecipeMockMvcTest.kt
package nl.rabobank.kotlinmovement.recipes.test.util

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import nl.rabobank.kotlinmovement.recipes.test.util.model.RecipeRequestTest
import nl.rabobank.kotlinmovement.recipes.test.util.model.RecipeResponseTest
import nl.rabobank.kotlinmovement.recipes.test.util.model.RecipesErrorResponseTest
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.HttpMethod
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.test.annotation.DirtiesContext
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.awaitBody
import org.springframework.web.reactive.function.client.awaitExchange

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
@AutoConfigureMockMvc
@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
class RecipeMockMvcTest {

    protected var objectMapper = jacksonObjectMapper()


    private var client: WebClient = WebClient.create("http://localhost:8080")


    @Throws(Exception::class)
    protected suspend fun setInitialState(): RecipeRequestTest {
        val ingredients = RecipeTestData.getDefaultIngredientRequests
        val initRecipe = RecipeRequestTest("Pizza", ingredients)
        createRecipe(initRecipe)
        return initRecipe
    }

    @Throws(Exception::class)
    protected suspend fun getRecipe(id: Long): RecipeResponseTest = mockMvcPerformRequest(
        HttpMethod.GET,
        "/recipes/$id",
        HttpStatus.OK
    )

    @Throws(Exception::class)
    protected suspend fun allRecipes(): Array<RecipeResponseTest> = mockMvcPerformRequest(
        HttpMethod.GET,
        "/recipes",
        HttpStatus.OK
    )

    //    @Throws(Exception::class)
//    protected fun updateRecipe(id: Long?, recipeRequest: RecipeRequestTest): RecipeResponseTest {
//        val requestBuilder = MockMvcRequestBuilders.put("/recipes/{id}", id)
//            .content(objectMapper.writeValueAsString(recipeRequest))
//            .contentType(MediaType.APPLICATION_JSON)
//        return mockMvcPerformRequest<RecipeResponseTest>(
//            requestBuilder,
//            MockMvcResultMatchers.status().isOk
//        )
//    }
//
//    @Throws(Exception::class)
//    protected fun badRequestCall(builder: MockHttpServletRequestBuilder): RecipesErrorResponseTest {
//        val requestBuilder = builder.contentType(MediaType.APPLICATION_JSON)
//        return mockMvcPerformRequest<RecipesErrorResponseTest>(
//            requestBuilder,
//            MockMvcResultMatchers.status().isBadRequest
//        )
//    }
//
    @Throws(Exception::class)
    protected suspend fun notFoundCall(httpMethod: HttpMethod, uri: String): RecipesErrorResponseTest =
        mockMvcPerformRequest(
            httpMethod,
            uri,
            HttpStatus.NOT_FOUND
        )


//    @Throws(Exception::class)
//    protected fun mockMvcPerformRequest(requestBuilder: MockHttpServletRequestBuilder, status: ResultMatcher) {
//        mockMvc
//            .perform(requestBuilder)
//            .andDo(MockMvcResultHandlers.print())
//            .andExpect(status)
//            .andReturn()
//            .response
//    }

    @Throws(Exception::class)
    private suspend fun createRecipe(recipe: RecipeRequestTest): RecipeResponseTest = mockMvcPerformRequest(
        HttpMethod.POST,
        "/recipes",
        HttpStatus.CREATED,
        objectMapper.writeValueAsString(recipe)
    )

    @Throws(Exception::class)
    private suspend inline fun <reified T : Any> mockMvcPerformRequest(
        method: HttpMethod,
        uri: String,
        status: HttpStatus,
        body: String? = null
    ): T {
        val i = client.method(method).uri(uri).contentType(MediaType.APPLICATION_JSON).accept(MediaType.APPLICATION_JSON)
        body?.let {
            i.bodyValue(it)
        }

        return i.awaitExchange {
            if (it.statusCode() == status) {
                it.awaitBody()
            } else {
                throw AssertionError("Unexpected status $status")
            }
        }
    }
}

